### PR TITLE
Service fabric new application cmdlets

### PR DESCRIPTION
**Pull request by a-santamaria**
## Description

Adding new cmdlets to manage appliaction and services:
    - New-AzServiceFabricApplication
    - New-AzServiceFabricApplicationType
    - New-AzServiceFabricApplicationTypeVersion
    - New-AzServiceFabricService
    - Update-AzServiceFabricApplication
    - Get-AzServiceFabricApplication
    - Get-AzServiceFabricApplicationType
    - Get-AzServiceFabricApplicationTypeVersion
    - Get-AzServiceFabricService
    - Remove-AzServiceFabricApplication
    - Remove-AzServiceFabricApplicationType
    - Remove-AzServiceFabricApplicationTypeVersion
    - Remove-AzServiceFabricServic
Upgrading Service Fabric SDK to version 1.2.0 which uses service fabric resource provider api-version 2019-03-01.

## Checklist

- [x] I have read the [_Submitting Changes_](https://github.com/Azure/azure-powershell/blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](https://github.com/Azure/azure-powershell/blob/master/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate [change log has been updated](https://github.com/Azure/azure-powershell/blob/master/CONTRIBUTING.md#updating-the-change-log)
- [x] The PR does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [x] For public API changes to cmdlets:
    - [x] the changes have gone through a [cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (Microsoft internal partners only) open issue: https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/336
    - [x] the cmdlet markdown files were [generated using the `platyPS` module](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/help-generation.md)

### CODE REVIEW

**Comment by isra-fel** [comment]
@a-santamaria , please mention me when the cmdlet design review is complete, thanks!

**Comment by a-santamaria** [comment]
@isra-fel the design review was accepted, thanks.

File: src/ServiceFabric/ServiceFabric/Commands/ProxyResourceCmdletBase.cs [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+namespace Microsoft.Azure.Commands.ServiceFabric.Commands
+{
+    public class ProxyResourceCmdletBase : ServiceFabricCmdletBase
+    {
+        /// <summary>
+        /// Cluster name
+        /// </summary>
+        [Parameter(Mandatory = true, Position = 1, ValueFromPipelineByPropertyName = true,
+            HelpMessage = "Specify the name of the cluster.")]
+        [ResourceGroupCompleter]
```
Comment (isra-fel) [inline src/ServiceFabric/ServiceFabric/Commands/ProxyResourceCmdletBase.cs]: same here

File: src/ServiceFabric/ServiceFabric/Commands/RemoveAzServiceFabricApplicationTypeVersion.cs [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        [ValidateNotNullOrEmpty()]
+        public override string ResourceGroupName { get; set; }
+
+        [Parameter(Mandatory = true, Position = 1, ValueFromPipelineByPropertyName = true,
+            HelpMessage = "Specify the name of the cluster.")]
+        [ResourceGroupCompleter]
+        [ValidateNotNullOrEmpty()]
+        public override string ClusterName { get; set; }
+
+        [Parameter(Mandatory = true, ValueFromPipeline = true,
```
Comment (isra-fel) [inline src/ServiceFabric/ServiceFabric/Commands/RemoveAzServiceFabricApplicationTypeVersion.cs]: same here about limiting `ValueFromPipeline`

File: src/ServiceFabric/ServiceFabric/Commands/UpdateAzServiceFabricApplication.cs [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+
+        [Parameter(Mandatory = false)]
+        public int? UpgradeTimeoutSec { get; set;}
+
+        #endregion
+
+        #region ApplicationHealthPolicy
+
+        [Parameter(Mandatory = false)]
+        public bool? ConsiderWarningAsError { get; set; }
```
Comment (isra-fel) [inline src/ServiceFabric/ServiceFabric/Commands/UpdateAzServiceFabricApplication.cs]: it should be a switch parameter, see https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/design-guidelines/parameter-best-practices.md#bool-vs-switchparameter

File: src/ServiceFabric/ServiceFabric/Commands/RemoveAzServiceFabricService.cs [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+        [ValidateNotNullOrEmpty()]
+        public override string ResourceGroupName { get; set; }
+
+        [Parameter(Mandatory = true, Position = 1, ParameterSetName = ByResourceGroup, ValueFromPipelineByPropertyName = true,
+            HelpMessage = "Specify the name of the cluster.")]
+        [ResourceNameCompleter("Microsoft.ServiceFabric/clusters", nameof(ResourceGroupName))]
+        [ValidateNotNullOrEmpty()]
+        public override string ClusterName { get; set; }
+
+        [Parameter(Mandatory = true, Position = 2, ParameterSetName = ByResourceGroup, ValueFromPipeline = true,
```
Comment (isra-fel) [inline src/ServiceFabric/ServiceFabric/Commands/RemoveAzServiceFabricService.cs]: Two `ValuneFromPipeline` in one cmdlet

File: tools/StaticAnalysis/Exceptions/Az.ServiceFabric/SignatureIssues.csv [hunk 5]
```diff
@@ -6,3 +6,8 @@
...
+"Microsoft.Azure.PowerShell.Cmdlets.ServiceFabric.dll","Microsoft.Azure.Commands.ServiceFabric.Commands.UpdateAzServiceFabricApplication","Update-AzServiceFabricApplication","1","8410","Parameter MaximumNodes of cmdlet Update-AzServiceFabricApplication does not follow the enforced naming convention of using a singular noun for a parameter name.","Consider using a singular noun for the parameter name."
+"Microsoft.Azure.PowerShell.Cmdlets.ServiceFabric.dll","Microsoft.Azure.Commands.ServiceFabric.Commands.UpdateAzServiceFabricApplication","Update-AzServiceFabricApplication","1","8410","Parameter DefaultServiceTypeMaxPercentUnhealthyServices of cmdlet Update-AzServiceFabricApplication does not follow the enforced naming convention of using a singular noun for a parameter name.","Consider using a singular noun for the parameter name."
```
Comment (isra-fel) [inline tools/StaticAnalysis/Exceptions/Az.ServiceFabric/SignatureIssues.csv]: I believe some of these parameters need rather to be renamed in signular than treated as exceptions. @markcowl what's your opinion?

**Comment by a-santamaria** [comment]
@isra-fel  @markcowl is this pr ready to be merged?

**Comment by isra-fel** [comment]
Sorry for late response @a-santamaria . Just please use singular form for paramters `MinimumNodes`, `MaximumNodes`, `DefaultServiceTypeMaxPercentUnhealthyServices` and `MaxPercentUnhealthyDeployedApplications`. Do not add them into exceptions. We have strict rules on using singular names for parameters. Details here: https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/design-guidelines/parameter-best-practices.md#singularity

**Comment by a-santamaria** [comment]
@isra-fel can we have these exceptions. I think removing the 's' at the end will change the meaning of these parameters and I don't think to have an 's' at the end necessarily means the parameters are plural. For example, DefaultServiceTypeMaxPercentUnhealthyServices the parameter that we are asking for is Maximum percent. Let me know if we can keep these otherewise I'll make the change

**Comment by cormacpayne** [comment]
@a-santamaria Hey Alfredo, how do the following parameter changes sound:

- `MinimumNodes` --> `MinimumNodeCount`
- `MaxiumNodes` --> `MaximumNodeCount`
- `DefaultServiceTypeMaxPercentUnhealthyServices` --> `DefaultServiceTypeUnhealthyServicesMaxPercent`
- `UnhealthyDeployApplicationsMaxPercent`

**Comment by a-santamaria** [comment]
thanks @cormacpayne, I change the parameters to the names you suggested.

File: src/ServiceFabric/ServiceFabric/Commands/UpdateAzServiceFabricApplication.cs [hunk 6]
```diff
@@ -141,12 +141,12 @@ public class UpdateAzServiceFabricApplication : ProxyResourceCmdletBase
         [Parameter(Mandatory = false, ParameterSetName = ByResourceGroup)]
         [Parameter(Mandatory = false, ParameterSetName = ByResourceId)]
         [ValidateRange(0, 100)]
-        public int DefaultServiceTypeMaxPercentUnhealthyServices { get; set; }
+        public int DefaultServiceTypeUnhealthyServicesMaxPercent { get; set; }
```
Comment (isra-fel) [inline src/ServiceFabric/ServiceFabric/Commands/UpdateAzServiceFabricApplication.cs]: Some of the parameters missed a proper `HelpMessage`. It helps a lot for understanding complex concepts like this one so we should add it.

File: src/ServiceFabric/ServiceFabric/Commands/NewAzServiceFabricService.cs [hunk 7]
```diff
@@ -0,0 +1,501 @@
...
+        public override string ClusterName { get; set; }
+
+        [Parameter(Mandatory = true, Position = 2, ParameterSetName = StatelessSingleton)]
+        [Parameter(Mandatory = true, Position = 2, ParameterSetName = StatelessUniformInt64)]
+        [Parameter(Mandatory = true, Position = 2, ParameterSetName = StatelessNamed)]
+        [Parameter(Mandatory = true, Position = 2, ParameterSetName = StatefulSingleton)]
+        [Parameter(Mandatory = true, Position = 2, ParameterSetName = StatefulUniformInt64)]
+        [Parameter(Mandatory = true, Position = 2, ParameterSetName = StatefulNamed)]
+        [ValidateNotNullOrEmpty()]
+        public string ApplicationName { get; set; }
```
Comment (isra-fel) [inline src/ServiceFabric/ServiceFabric/Commands/NewAzServiceFabricService.cs]: Please also add `HelpMessage` in this file

File: src/ServiceFabric/ServiceFabric/Commands/UpdateAzServiceFabricApplication.cs (hunk 6 above)
Comment (a-santamaria) [reply src/ServiceFabric/ServiceFabric/Commands/UpdateAzServiceFabricApplication.cs]: yes, I had forgotten. done

**Comment by a-santamaria** [comment]
> @a-santamaria
> 
> 1. Please add `ResourceIdCompleterAttribute` to the `ResourceId` parameters
> 2. Please clean up the commits [(details)](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md#best-practices-for-keeping-commits-clean)
> 
> And your PR would be ready to merge.
> Thanks

@isra-fel done, thanks

**Comment by isra-fel** [comment]
@a-santamaria 
There are still too many commits... and we wouldn't want some of them to appear in the history, like "fixing pr comments", right?
So please squash them into one commit, so that I can merge it. Thanks!

**Comment by a-santamaria** [comment]
@isra-fel I thought you would merge-squash them into one. Should I squash all of them in one commit?

**Comment by a-santamaria** [comment]
@isra-fel I squash the last 13 commits into one

**Pull request merged**